feat: <Button width="full" align="start|center|end">

gnapse opened the pull request:
## Short description

During my work in the new Task details view in Todoist, I've noticed the need for buttons to be optionally expanding their width instead of growing as little as needed to have their label fit. Also, when in this mode, we want to customize the label alignment as well.

<img width="521" alt="image" src="https://user-images.githubusercontent.com/15199/163079649-35d73ece-2d52-4e2d-a325-45b9fc924034.png">

I recommend looking at the diff with whitespace changes hidden.

## Test plan

Using storybook (`npm run storybook`), do the following:

- [x] The most important thing is to validate that buttons not using this new feature are 100% unaffected visually (I recommend using [the button playground story](http://localhost:6006/?path=/docs/design-system-button--playground)).
- [x] Play with the new full-width button in [the button full-width story](http://localhost:6006/?path=/docs/design-system-button--full-width). Play with different alignment values (left, center and right), and also change some other attributes (variant, size, long vs. short labels, etc.) See that it all looks good.

## PR Checklist

<!--
Feel free to leave unchecked or remove the lines that are not applicable.
-->

-   [x] Added tests for bugs / new features
-   [x] Updated docs (storybooks, readme)
-   [x] Executed `npm run validate` and made sure no errors / warnings were shown
-   [x] Described changes in `CHANGELOG.md`
-   [ ] ~~Bumped version in `package.json` and `package-lock.json` (`npm --no-git-tag-version version <major|minor|patch>`) [ref](https://docs.npmjs.com/cli/v6/commands/npm-version)~~
-   [ ] ~~Updated all static build artifacts (`npm run build-all`)~~

## Versioning

New minor release.

frankieyan commented on src/new-components/base-button/base-button.tsx:
```diff
@@ -111,6 +117,8 @@ export const BaseButton = polymorphicComponent<'div', BaseButtonProps>(function
                 styles[`variant-${variant}`],
                 styles[`tone-${tone}`],
                 styles[`size-${size}`],
+                width !== 'auto' && icon == null && align != null ? styles[`align-${align}`] : null,
```
Since `align` is only meant to be used with the `width="full"` prop, it might be helpful to guard against them with something like this (haven't tried if it works though):

```TS
type AlignmentProps = {
	width: 'full'
	align?: 'left' | 'center' | 'right'
}

type AutoWidthProps = {
	width?: 'auto'
	align?: never
}

type LabelledButtonProps = { ... } & (AlignmentProps | AutoWidthProps)
```

gnapse replied:
I thought about it, and as you know, we already do something along those lines to take care of the icon-only vs. the full-visual-text-label cases, as these have even more striking differences. I was hesitant, though, to introduce this level of complexity in the types. But maybe is something that the library can contain.

Another alternative that I thought was to not introduce the `width` prop at all. And make the button be full width whenever the `align` property is provided. But this on the other hand seemed a bit obscure.

Thoughts?

frankieyan replied:
Yeah I don't think it should be omitted. It's not obvious `<Button align="center">` should be expanded just by looking at it 🤔 I think the explicitness of having `width="full"` in there would be more preferable.

gnapse replied:
That's what I thought too.

I'll tinker the types a bit to see if I can achieve this without complicating things too much.

gnapse replied:
Your suggestion seems to be working well. I'll validate a few things, but it looks like we'll make it.

frankieyan commented on src/new-components/button-link/button-link.stories.mdx:
```diff
@@ -478,6 +478,92 @@ export function LoadingButtonLink(props) {
+        }}
+        name="Full-width"
+    >
+        {FullWidthTemplate.bind({})}
```
I guess we can only pick between having controls vs having code snippets. Right now this shows up like this when you click on show code:

![image](https://user-images.githubusercontent.com/8531248/163137757-16387687-9443-4883-8579-986840ca9f16.png)

gnapse replied:
Yup, something we need to figure out with storybook. Perhaps it has bothered someone else already and there's a way around it. Not an issue introduced here, as all the other examples using templates are the same.

frankieyan commented on src/new-components/button-link/button-link.test.tsx:
```diff
@@ -217,6 +217,27 @@ describe('ButtonLink', () => {
+
+    it('ignores align when width is not full', () => {
+        render(
+            <ButtonLink href="/" variant="primary" align="right">
```
Would be great if this can trigger `@ts-expect-error`

gnapse replied:
Done

frankieyan commented on src/new-components/button/button.stories.mdx:
```diff
@@ -467,9 +467,95 @@ export function LoadingButton(props) {
+                    'Mark as done',
+                    'Yes, cancel my subscription',
+                    'Click me <em>now</em>',
+                    'If you click me now, youʼll take away the biggest part of me',
```
😂

frankieyan commented on src/new-components/button-link/button-link.stories.mdx:
```diff
@@ -478,6 +478,92 @@ export function LoadingButtonLink(props) {
+            endIcon: { control: false },
+            icon: { control: false },
+            tooltip: { control: false },
+            exceptionallySetClassName: { control: false },
```
Can we add `{ control: false }` for `align` and `width` as well? Right now they render controls that don't do anything.

gnapse replied:
Good point.

gnapse replied:
Done

frankieyan reviewed (approved):
Looks good, just a minor comment to see if it's possible to guard the align and width props against each other

gnapse commented:
@frankieyan, in addition to adding your suggestion about the stricter types, I realized that the concept of `width` is already implemented in `Box`, and a button renders a `Box`, so I changed it so that we need no extra CSS for the width in `base-button.module.css`.

Moreover, I realized that I could expand this to not be only about `width="full"`, but any of the predefined box width values will also work. Hence, I added that change as well.

Since this is a non-trivial change, I'll wait for one last sanity check from you before I merge this. There's no hurry, and I'd rather be cautious with library changes like this.

frankieyan commented:
@gnapse my only comment is that I think the widths seem a little large for buttons (e.g. `small` is 400px). Even on the `<Box>` component, the `width` prop isn't getting too much use other than for its `full` value. That said, we could always introduce this now and come up with different values with the design team later, which shouldn't be too hard if we don't have a lot of use cases right now.

pauloslund commented on src/new-components/base-button/base-button.module.css:
```diff
@@ -89,6 +83,26 @@
+    text-overflow: ellipsis;
+    white-space: nowrap;
+    overflow: hidden;
+    width: 100%;
```
@gnapse I believe this will introduce a breaking change to the Twist login page when Reactist is updated to this version:
| Before | After |
| --- | --- |
| <img width="386" alt="image" src="https://user-images.githubusercontent.com/1051164/168882438-2fcc424e-8fb2-4e32-9f32-c1a074593f43.png"> | <img width="386" alt="image" src="https://user-images.githubusercontent.com/1051164/168882514-347088f0-0853-4050-a29d-cdaf0a4b11d7.png"> |

The only way around it seems to be adding custom CSS. Would it make sense for `width: 100%;` to only be applied when width is `full`?

gnapse replied:
Indeed, if that fixes it, it makes sense that the label does not need to stretch in all situations. Does that fix it?

pauloslund replied:
Yeah it seems to fix it. I'll create a PR for it!
